File: Basic/Python/1015.py
'''
@Author: kurumi
@Date: 2025-12-21
@Description: 1015 德才论
'''
from typing import List

class Student:
    def __init__(self, id: str, moral: int, ability: int):
        self.id = id
        self.moral = moral
        self.ability = ability
        self.total = moral + ability

    def __str__(self):
        return f'{self.id} {self.moral} {self.ability}'

    def isAdmitted(self, l: int) -> bool:
        return self.moral >= l and self.ability >= l

    def isMoralAndAbility(self, h: int) -> bool:
        return self.moral >= h and self.ability >= h

    def isMoralMoreThanAbilityH(self, h: int) -> bool:
        return self.moral >= h and self.ability < h

    def isMoralMoreThanAbilityL(self, h: int) -> bool:
        return self.moral < h and self.ability < h and self.moral >= self.ability

    def compare(self, stu) -> bool:
        if self.total == stu.total:
            return self.moral > stu.moral if self.moral != stu.moral else self.id < stu.id
        return self.total > stu.total

def main():
    n, l, h = map(int, input().split())
    moralAndAbilityStudents: List[Student] = list()
    moralMoreThanAbilityHStudents: List[Student] = list()
    moralMoreThanAbilityLStudents: List[Student] = list()
    otherStudents: List[Student] = list()
    cnt: int = 0
    for i in range(n):
        id, moral, ability = input().split()
        stu: Student = Student(id, int(moral), int(ability))
        if stu.isAdmitted(l):
            cnt += 1
            if stu.isMoralAndAbility(h):
                moralAndAbilityStudents.append(stu)
            elif stu.isMoralMoreThanAbilityH(h):
                moralMoreThanAbilityHStudents.append(stu)
            elif stu.isMoralMoreThanAbilityL(h):
                moralMoreThanAbilityLStudents.append(stu)
            else:
                otherStudents.append(stu)
    moralAndAbilityStudents.sort(key = lambda stu: (-stu.total, -stu.moral, stu.id))
    moralMoreThanAbilityHStudents.sort(key = lambda stu: (-stu.total, -stu.moral, stu.id))
    moralMoreThanAbilityLStudents.sort(key = lambda stu: (-stu.total, -stu.moral, stu.id))
    otherStudents.sort(key = lambda stu: (-stu.total, -stu.moral, stu.id))

    print(cnt)
    for stu in moralAndAbilityStudents:
        print(stu)
    for stu in moralMoreThanAbilityHStudents:
        print(stu)
    for stu in moralMoreThanAbilityLStudents:
        print(stu)
    for stu in otherStudents:
        print(stu)

if __name__ == '__main__':
    main()

File: Basic/JS/1015.js
/**
 * @Author kurumi
 * @Date 2025-12-21
 * @Description 1015 德才论
 */
const rl = require("readline").createInterface({input: process.stdin});
const iter = rl[Symbol.asyncIterator]();
const readline = async () => (await iter.next()).value;

class Student {
  constructor(id, moral, ability) {
    this.id = id;
    this.moral = moral;
    this.ability = ability;
    this.total = moral + ability;
  }
  
  isAdmitted(l) {
    return this.moral >= l && this.ability >= l;
  }
  
  isMoralAndAbility(h) {
    return this.moral >= h && this.ability >= h;
  }
  
  isMoralMoreThanAbilityH(h) {
    return this.moral >= h && this.ability < h;
  }
  
  isMoralMoreThanAbilityL(h) {
    return this.moral < h && this.ability < h && this.moral >= this.ability;
  }
  
  compare(stu) {
    if (this.total == stu.total) {
      if (this.moral != stu.moral) return stu.moral - this.moral;
      return this.id.localeCompare(stu.id);
    }
    return stu.total - this.total;
  }
  
  toString() {
    return `${this.id} ${this.moral} ${this.ability}`;
  }
}

void (async function() {
  const [n, l, h] = (await readline()).split(" ").map(Number);
  const moralAndAbilityStudents = new Array();
  const moralMoreThanAbilityHStudents = new Array();
  const moralMoreThanAbilityLStudents = new Array();
  const otherStudents = new Array();
  for (let i = 0; i < n; i++) {
    const [id, moral, ability] = (await readline()).split(" ");
    const student = new Student(id, parseInt(moral), parseInt(ability));
    if (student.isAdmitted(l)) {
      if (student.isMoralAndAbility(h)) moralAndAbilityStudents.push(student);
      else if (student.isMoralMoreThanAbilityH(h)) moralMoreThanAbilityHStudents.push(student);
      else if (student.isMoralMoreThanAbilityL(h)) moralMoreThanAbilityLStudents.push(student);
      else otherStudents.push(student);
    }
  }
  moralAndAbilityStudents.sort((stu1, stu2) => stu1.compare(stu2));
  moralMoreThanAbilityHStudents.sort((stu1, stu2) => stu1.compare(stu2));
  moralMoreThanAbilityLStudents.sort((stu1, stu2) => stu1.compare(stu2));
  otherStudents.sort((stu1, stu2) => stu1.compare(stu2));
  
  console.log(moralAndAbilityStudents.length + moralMoreThanAbilityHStudents.length + moralMoreThanAbilityLStudents.length + otherStudents.length);
  moralAndAbilityStudents.forEach(stu => console.log(stu.toString()));
  moralMoreThanAbilityHStudents.forEach(stu => console.log(stu.toString()));
  moralMoreThanAbilityLStudents.forEach(stu => console.log(stu.toString()));
  otherStudents.forEach(stu => console.log(stu.toString()));
})();